projects/plugins/protect/src/class-scan-history.php

```diff
@@ -0,0 +1,383 @@
+		// TODO: Sanitize $history.
+		update_option( static::OPTION_NAME, maybe_serialize( $history ) );
+		update_option( static::OPTION_TIMESTAMP_NAME, time() + static::OPTION_EXPIRES_AFTER );
+	}
+
+	/**
+	 * Gets the current history of the Jetpack Protect checks
+	 *
+	 * @param bool  $refresh_from_wpcom Refresh the local plan and history cache from wpcom.
+	 * @param array $filter The filter to apply to the data.
+	 * @return History_Model|bool
+	 */
+	public static function get_scan_history( $refresh_from_wpcom = false, $filter = null ) {
+		$has_required_plan = Plan::has_required_plan();
+		if ( ! $has_required_plan ) {
+			return false;
+		}
+
+		if ( self::$history !== null ) {
+			return self::$history;
+		}
+
+		if ( $refresh_from_wpcom || ! self::should_use_cache() || self::is_cache_expired() ) {
+			$history = self::fetch_from_api();
+		} else {
+			$history = self::get_from_options();
+		}
+
+		if ( is_wp_error( $history ) ) {
+			$history = new History_Model(
+				array(
+					'error'         => true,
+					'error_code'    => $history->get_error_code(),
+					'error_message' => $history->get_error_message(),
+				)
+			);
+		} else {
+			$history = self::normalize_api_data( $history, $filter );
+		}
+
+		self::$history = $history;
+		return $history;
+	}
+
+	/**
+	 * Gets the Scan API endpoint
+	 *
+	 * @return WP_Error|string
+	 */
+	public static function get_api_url() {
+		$blog_id      = Jetpack_Options::get_option( 'id' );
+		$is_connected = ( new Connection_Manager() )->is_connected();
+
+		if ( ! $blog_id || ! $is_connected ) {
+			return new WP_Error( 'site_not_connected' );
+		}
+
+		$api_url = sprintf( self::SCAN_HISTORY_API_BASE, $blog_id );
+
+		return $api_url;
+	}
+
+	/**
+	 * Fetches the history data from the Scan API
+	 *
+	 * @return WP_Error|array
+	 */
+	public static function fetch_from_api() {
+		$api_url = self::get_api_url();
+		if ( is_wp_error( $api_url ) ) {
+			return $api_url;
+		}
+
+		$response = Client::wpcom_json_api_request_as_blog(
+			$api_url,
+			'2',
+			array( 'method' => 'GET' ),
+			null,
+			'wpcom'
+		);
+
+		$response_code = wp_remote_retrieve_response_code( $response );
+
+		if ( is_wp_error( $response ) || 200 !== $response_code || empty( $response['body'] ) ) {
+			return new WP_Error( 'failed_fetching_status', 'Failed to fetch Scan history from the server', array( 'status' => $response_code ) );
+		}
+
+		$body               = json_decode( wp_remote_retrieve_body( $response ) );
+		$body->last_checked = ( new \DateTime() )->format( 'Y-m-d H:i:s' );
+		self::update_history_option( $body );
+
+		return $body;
+	}
+
+	/**
+	 * Normalize API Data
+	 * Formats the payload from the Scan API into an instance of History_Model.
+	 *
+	 * @param object $scan_data The data returned by the scan API.
+	 * @param array  $filter    The filter to apply to the data.
+	 * @return History_Model
+	 */
+	private static function normalize_api_data( $scan_data, $filter ) {
```

Comment:
Opinionated: using early returns in this function might improve readability.

Reply:
I'm realizing there are some areas for improvement/optimization here, I'll look into that and touch base with you on how best to include this all in the new approach.